platform/src/main/java/org/hillview/main/BatchLogAnalysis.java

```diff
@@ -0,0 +1,188 @@
+        DataRange dataRange = table1.blockingSketch(rangeSketch);
+        DoubleHistogramBuckets bucketsTimestamp = new DoubleHistogramBuckets(dataRange.min, dataRange.max, numOfTimestampBuckets);
+
+        SampleDistinctElementsSketch sampleSketch = new SampleDistinctElementsSketch("errorCode", 0, 200);
+        MinKSet<String> samples = table1.blockingSketch(sampleSketch);
+        JsonList<String> leftBoundaries = samples.getLeftBoundaries(200);
+        StringHistogramBuckets bucketsErrorCode = new StringHistogramBuckets(leftBoundaries.toArray(new String[0]));
+
+        HeatmapSketch heatmapSketch = new HeatmapSketch(bucketsTimestamp, bucketsErrorCode, "Timestamp", "errorCode", 1.0, 0);
+        Heatmap heatmap = table1.blockingSketch(heatmapSketch);
+        HeatmapData heatmapData = new HeatmapData();
+        int numOfBucketsD1 = heatmap.getNumOfBucketsD1();
+        int numOfBucketsD2 = heatmap.getNumOfBucketsD2();
+        heatmapData.matrix = new long[numOfBucketsD1][numOfBucketsD2];
+        for (int i = 0; i<numOfBucketsD1; i++){
+            for (int j = 0; j<numOfBucketsD2; j++){
+                heatmapData.matrix[i][j] = heatmap.getCount(i, j);
+            }
+        }
+        heatmapData.errorCodeLabels = leftBoundaries;
+        for (int x = 0; x < numOfTimestampBuckets; x++){
+            double time = dataRange.min + x * (dataRange.max - dataRange.min) / numOfTimestampBuckets;
+            Instant instantTime = Converters.toDate(time);
+            String stringDate = Converters.toString(instantTime);
+            heatmapData.timeLabels.add(stringDate);
+        }
+        return heatmapData;
+    }
+
+    private static void saveHeatmapToFile(HeatmapData heatmapData, String filepath) {
+        try (PrintWriter writer = new PrintWriter(new File(filepath))) {
+            StringBuilder sb = new StringBuilder();
+            sb.append("Timestamp");
+            sb.append(',');
+            sb.append("errorCode");
+            sb.append(',');
+            sb.append("Count");
+            sb.append('\n');
+            for (int x = 0; x < heatmapData.matrix.length; x++) {
+                for (int y = 0; y < heatmapData.matrix[0].length; y++) {
+                    if (heatmapData.matrix[x][y] >= 0) {        // keep zeros or not
+                        sb.append(heatmapData.timeLabels.get(x));
+                        sb.append(',');
+                        sb.append(heatmapData.errorCodeLabels.get(y).replaceAll(",",""));
+                        sb.append(',');
+                        sb.append(heatmapData.matrix[x][y]);
+                        sb.append('\n');
+                    }
+                }
+            }
+            writer.write(sb.toString());
+        } catch (FileNotFoundException e) {
+            System.out.println(e.getMessage());
+        }
+    }
+
+    public static void getBugHeatmaps(FileSetDescription desc, LocalDataSet<Empty> local, String logDir, String figDir, int numOfTimestampBuckets) {
```

Comment:
Please add a JavaDoc comment to this method describing each of the arguments.
I don't think you need the local argument, you can just create it locally.